src/NHibernate/Type/AbstractDateTimeType.cs

```diff
@@ -0,0 +1,160 @@
+			Get(rs, rs.GetOrdinal(name), session);
+
+		/// <summary>
+		/// Get the <see cref="DateTime" /> in the <see cref="DbDataReader"/> for the Property.
+		/// </summary>
+		/// <param name="rs">The <see cref="DbDataReader"/> that contains the value.</param>
+		/// <param name="index">The index of the field to get the value from.</param>
+		/// <param name="session">The session for which the operation is done.</param>
+		/// <returns>An object with the value from the database.</returns>
+		protected virtual DateTime GetDateTime(DbDataReader rs, int index, ISessionImplementor session)
+		{
+			try
+			{
+				return AdjustDateTime(Convert.ToDateTime(rs[index]));
+			}
+			catch (Exception ex)
+			{
+				throw new FormatException($"Input string '{rs[index]}' was not in the correct format.", ex);
+			}
+		}
+
+		/// <inheritdoc />
+		public override void Set(DbCommand st, object value, int index, ISessionImplementor session) =>
+			st.Parameters[index].Value = GetDateTimeToSet(value, session);
+
+		/// <summary>
+		/// Get the <see cref="DateTime" /> from the value to set in a parameter.
+		/// </summary>
+		/// <param name="value">The original value.</param>
+		/// <param name="session">The session for which the operation is done.</param>
+		/// <returns>The <see cref="DateTime" /> to set as parameter value.</returns>
+		protected virtual DateTime GetDateTimeToSet(object value, ISessionImplementor session) =>
+			AdjustDateTime((DateTime) value);
+
+		#region IVersionType Members
+
+		/// <inheritdoc />
+		public object Next(object current, ISessionImplementor session) =>
+			Seed(session);
+
+		/// <summary>
+		/// Round a <see cref="DateTime"/> according to specified resolution.
+		/// </summary>
+		/// <param name="value">The value to round.</param>
+		/// <param name="resolution">The resolution in ticks (100ns).</param>
+		/// <returns>A rounded <see cref="DateTime"/>.</returns>
+		public static DateTime Round(DateTime value, long resolution) =>
+			value.AddTicks(-(value.Ticks % resolution));
+
+		/// <inheritdoc />
+		public virtual object Seed(ISessionImplementor session) =>
+			session == null ? Now : Round(Now, session.Factory.Dialect.TimestampResolutionInTicks);
+
+		/// <inheritdoc />
+		public virtual IComparer Comparator => Comparer<DateTime>.Default;
+
+		#endregion
+
+		/// <summary>
+		/// Compares two <see cref="DateTime" /> object and also compare its Kind if needed, which is not used by the
+		/// .Net Framework <see cref="DateTime.Equals(object)"/> implementation.
+		/// </summary>
+		/// <param name="x">The first date time to compare.</param>
+		/// <param name="y">The second date time to compare.</param>
+		/// <returns><see langword="true" /> if they are equals, <see langword="false" /> otherwise.</returns>
+		public override bool IsEqual(object x, object y) =>
+			base.IsEqual(x, y) &&
+				(Kind == DateTimeKind.Unspecified || x == null || ((DateTime) x).Kind == ((DateTime) y).Kind);
```

Comment:
`GetHashCode` is not overridden but that should not cause issues: this `IsEqual` further restrict the equality, meaning the contract will not be broken. Two equal object will still have the same hashcode. `GetHashCode` is overridden for thos date types only when the equality is relaxed, like in `DateType` or the new `DateTimeNoMsType` (old `DateTimeType`).